core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/AvroUpsertMonitor.scala

```diff
@@ -0,0 +1,234 @@
+    durations.max
+  }
+
+  override def receive = {
+    case StartMonitorPass =>
+      // start the process by pulling a message and sending it back to self
+      pubSubDao.pullMessages(pubSubSubscriptionName, 1).map(_.headOption) pipeTo self
+
+    case Some(message: PubSubMessage) =>
+      //we received a message, so we will parse it and try to upsert it
+      logger.info(s"received avro upsert message: $message")
+      val (jobId, file) = parseMessage(message)
+
+      file match {
+        case "upsert.json" => initAvroUpsert(jobId, message.ackId).map(_ => ImportComplete) pipeTo self
+        case _ => acknowledgeMessage(message.ackId).map(_ => None) pipeTo self //some other file (i.e. success/failure log, metadata.json)
+      }
+
+    case None =>
+      // there was no message so wait and try again
+      val nextTime = org.broadinstitute.dsde.workbench.util.addJitter(pollInterval, pollIntervalJitter)
+      system.scheduler.scheduleOnce(nextTime, self, StartMonitorPass)
+
+    case ImportComplete => self ! None
+
+    case Status.Failure(t) => throw t
+
+    case ReceiveTimeout => throw new WorkbenchException("AvroUpsertMonitorActor has received no messages for too long")
+
+    case x =>
+      logger.info(s"unhandled $x")
+      self ! None
+  }
+
+  private def initAvroUpsert(jobId: String, ackId: String): Future[Unit] = {
+    //start this Future asap because it can be long-running
+    val readUpsertObjectFuture = readUpsertObject(s"$jobId/upsert.json")
+
+    for {
+      avroMetadataJson <- readMetadataObject(s"$jobId/metadata.json")
+      //ack the response after we load the json into memory. pro: don't have to worry about ack timeouts for long operations, con: if someone restarts rawls here the uspert is lost
+      petSAJson <- samDAO.getPetServiceAccountKeyForUser(avroMetadataJson.namespace, RawlsUserEmail(avroMetadataJson.userEmail))
+      petUserInfo <- googleServicesDAO.getUserInfoUsingJson(petSAJson)
+      avroUpsertJson <- readUpsertObjectFuture
+      ackResponse <- acknowledgeMessage(ackId)
+      upsertResults <-
+        avroUpsertJson.grouped(batchSize).toList.traverse { upsertBatch =>
+          IO.fromFuture(IO(workspaceService.apply(petUserInfo).batchUpdateEntities(WorkspaceName(avroMetadataJson.namespace, avroMetadataJson.name), upsertBatch, true)))
+        }.unsafeToFuture
+    } yield {
+      logger.info(s"completed Avro upsert job ${avroMetadataJson.jobId} for user: ${avroMetadataJson.userEmail} with size ${avroUpsertJson.size} entities")
+      ()
+    }
+  }
+
+  private def acknowledgeMessage(ackId: String) =
+    pubSubDao.acknowledgeMessagesById(pubSubSubscriptionName, Seq(ackId))
+
+  private def parseMessage(message: PubSubMessage) = {
+    message.contents.parseJson.asJsObject.getFields("name").head.compactPrint match {
+      case objectIdPattern(jobId, file) => (jobId, file)
+      case m => throw new Exception(s"unable to parse message $m")
+    }
+  }
```

Comment:
talking out loud, and NOT a blocker for this PR ... I'm very slightly nervous about relying on `compactPrint` behavior. If the libraries ever change how they print out json to a String, this could break. We could skip the regex and get right at the data, something like:
```
  private def parseMessage(message: PubSubMessage) = {
    message.contents.parseJson.asJsObject.getFields("name").head match {
      case jss:JsString =>
        val pathParts = jss.value.split('/')
        if (pathParts.size == 2) (pathParts(0), pathParts(1)) else throw new Exception(s"message has wrong parts: $jss")
      case m => throw new Exception(s"unable to parse message $m")
    }
  }
```
again, no need to change anything for this PR.